core/utils/historied-data/src/lib.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! History driven data storage.
+//! Useful to store information with history
+//! on a per item basis.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub mod linear;
+
+#[derive(Debug, Clone, PartialEq, Copy)]
+/// State of a data.
+pub enum State {
+	/// Data is under change and can still be dropped.
+	Pending,
+	/// Data is under change and can still be dropped.
+	/// This also mark the start of a transaction.
+	TxPending,
```

Comment:
I think when we drop the prospective stuff from the overlayedchanges, we probably can merge `Pending` and `TxPending`?